sys/arm64/arm64/elf_machdep.c

```diff
@@ -476,11 +488,32 @@ elf_reloc_internal(linker_file_t lf, char *relocbase, const void *data,
 			*where = elf_relocaddr(lf, (Elf_Addr)relocbase + addend);
 #if __has_feature(capabilities)
 		else if (rtype == R_MORELLO_RELATIVE) {
-			cap = build_cap_from_fragment(where,
-			    (Elf_Addr)relocbase, addend,
-			    (__cheri_tocap void * __capability)relocbase,
-			    (__cheri_tocap void * __capability)relocbase);
-			*(uintcap_t *)(void *)where = cap;
+			void * __capability base;
+			Elf_Addr addr1, size;
+			uint8_t perms;
+
+			decode_fragment(where, (Elf_Addr)relocbase, &addr,
+			    &size, &perms);
+
+			/*
+			 * Handle relocations against magic DPCPU and VNET
+			 * symbols: the address is transformed to refer to a
+			 * segment in the base kernel's DPCPU/VNET segments.
+			 * In this case we must use the kernel's base
+			 * capability.
+			 */
+			addr1 = elf_relocaddr(lf, addr + addend);
+			if (addr + addend == addr1) {
+				addr1 = addr;
+				base = (__cheri_tocap void * __capability)
+				    relocbase;
+			} else {
+				addend = 0;
+				base = (__cheri_tocap void * __capability)
+				    linker_kernel_file->address;
+			}
```

Comment:
Given what elf_relocaddr does, I think the simplest thing to do that works is just this?

```suggestion
			addr = elf_relocaddr(lf, addr + addend) - addend;
```

Though arguably if that's not equivalent to the following then the input was very dodgy:

```suggestion
			addr = elf_relocaddr(lf, addr);
```

Unlike normal R_FOO_RELATIVE, where the addend is the offset from the entire start of the DSO's mapping, for R_MORELLO_RELATIVE it's the offset from the capability's base address, which really should reside within the regions if and only if the address as a whole does.

Reply:
> for R_MORELLO_RELATIVE it's the offset from the capability's base address, which really should reside within the regions if and only if the address as a whole does.

The last part wasn't really obvious to me from the Morello ABI docs, but it'd be pretty surprising if it weren't true.